Description: Make grub-probe work with symlinks under /dev/mapper
 Previously, real block devices were more common here, but the Linux world
 seems to be moving towards symlinks.  GRUB shouldn't get caught in the
 middle and should just work with both.
Author: Colin Watson <cjwatson@ubuntu.com>
Bug-Debian: http://bugs.debian.org/550704
Forwarded: http://lists.gnu.org/archive/html/grub-devel/2010-05/msg00196.html
Applied-Upstream: http://bazaar.launchpad.net/~vcs-imports/grub/grub2-bzr/revision/2403
Last-Update: 2010-05-28

Index: b/kern/emu/getroot.c
===================================================================
--- a/kern/emu/getroot.c
+++ b/kern/emu/getroot.c
@@ -126,9 +126,20 @@
 	/* Ignore any error.  */
 	continue;
 
-      if (S_ISLNK (st.st_mode))
-	/* Don't follow symbolic links.  */
+      if (S_ISLNK (st.st_mode)) {
+#ifdef __linux__
+	if (strcmp (dir, "mapper") == 0) {
+	  /* Follow symbolic links under /dev/mapper/; the canonical name
+	     may be something like /dev/dm-0, but the names under
+	     /dev/mapper/ are more human-readable and so we prefer them if
+	     we can get them.  */
+	  if (stat (ent->d_name, &st) < 0)
+	    continue;
+	} else
+#endif /* __linux__ */
+	/* Don't follow other symbolic links.  */
 	continue;
+      }
 
       if (S_ISDIR (st.st_mode))
 	{
